components/director/internal/domain/application/service.go

```diff
@@ -794,6 +797,14 @@ func (s *service) Merge(ctx context.Context, destID, srcID string) (*model.Appli
 		return nil, errors.Errorf("Application templates are not the same. Destination app template: %s. Source app template: %s", destTemplateID, srcTemplateID)
 	}
 
+	appTemplateLabels, err := s.labelRepo.ListForObject(ctx, appTenant, model.AppTemplateLabelableObject, srcTemplateID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while getting labels for app template with id %s", srcTemplateID)
+	}
+
+	if _, exists := appTemplateLabels[s.selfRegisterDistinguishLabelKey]; exists {
+		return nil, errors.Errorf("app template: %s has label %s", srcTemplateID, s.selfRegisterDistinguishLabelKey)
```

Comment:
Before returning this error add logger(Log.C(ctx).Infof..) which says that the source app template has a self-reg label and should not be merged. Because now it is not clear why we are returning an error if we have such label.